### DISCUSSION TITLE
Design Patterns in the year 2022

### DISCUSSION THREAD

**Discussion by guettli** [guettli/programming-guidelines#17] [General]
@zxenonx

I think design patterns are overrated. Today, you never start from scratch. You eigther have a database (for example potsgresql) or you have an API. The services you work with define their usage pattern.

A lot of design patterns are about native GUI applications. And today your mostly write http request handlers. At least in my context. The http request comes in, your code contacts a DB, API or uses a library, and then returns the result. 

In this context you hardly need design patterns. I read a book about design patterns 20 years ago, and I have never used it. Maybe this pattern of stateless http request handling is a design pattern. I don't know.

**Comment by zxenonx** [comment]
> 

Yes, I quite agree with what you said.

Most of the tools we are using today are quite opinionated.

However, I think design patterns can be useful not for structuring the whole codebase but for example in a certain part of the codebase to tackle a specific issue.

What do you think?

**Reply by guettli** [reply]
I think it is important to avoid too abstract blu-blu-bla-bla. It is important to focus on the customer and on simple solutions.

AFAIK there is a book about Design Patterns in Python.

I read a book about design patterns (20 years ago) which used Java for the examples.

Many patterns where not needed in Python, because Python had a rich standard-library even in the year 2002.

Since then I live happily without design patterns.

Tests are important. If you have tested most of your code, then you can refactor it, and you can be sure that the user/customer won't notice your refactoring.

**Reply by zxenonx** [reply]
> Many patterns where not needed in Python, because Python had a rich standard-library even in the year 2002.

Very interesting.

In fact when I read some articles about design patterns, it seems to me that most of them arise from some limitations the languages then in vogue (Smalltalk, Objective-C).

I don't know if it's correct.

**Reply by zxenonx** [reply]
With all the experience you have gained, sir, I am already excited as I know I'm going to learn a lot from you.

I'm really lucky to have stumbled upon your notes.
Thank you very much.
